### ISSUE TITLE
Listener can't call stop() ?

### ISSUE THREAD

**Issue by rudiedirkx** [salsify/jsonstreamingparser#50]
I might be missing something here, but it looks like the listener class can't call the parser's `stop()`, because it doesn't have a parser object. The perfect place to introduce them would be in `Parser::__construct()`, where both objects are ready for use.

`setParser(Parser $parser)` would be optimal, but that would break bc for classes implementing the interface directly.

What I did to 'fix' this, is throw and handle a `ParsingStoppedException`, but that seems silly since there is a `stop()` method...

**Comment by gonzofy** [comment]
@rudiedirkx good point. One thing we could do is to add a `setListener()` method to the parser, this would enable you to create the parser, then create a listener, passing in the parser in the constructor as you suggest, and THEN have the parser `setListener(self)` to close the loop.

What do you think?

**Comment by rudiedirkx** [comment]
Not sure how you mean... You have to create the listener before the parser, right?

    $listener = new Listener;
    $parser = new Parser($stream, $listener);

or something like that? So the one point they surely meet is in the Parser constructor.

The other way seems more correct to me, because the Listener is completely custom, and wants to know the parser, so part of being a Listener, is storing a Parser.

Practically it would even work without interfacing. `Parser::__construct()` could just add itself to the listener like `$listener->parser = $this;`, but that's not very correct.

**Comment by gonzofy** [comment]
The general programming pattern is that you set listeners on an object that emits events, not the other way around as you're suggesting.

So what I'm saying is that if we break up the setting of the listener from the constructor and allow it to be set afterwards you can achieve what you're looking for.

1. Create parser.
2. Create listener. Pass in parser.
3. `setListener()` on parser so parser can emit events effectively.

Make sense?

**Comment by rudiedirkx** [comment]
Makes sense, but that's not how it works, and it's not how it can work with bc. The parser expects 1 listener, never 0 or 2. Requiring and using always 1 is super efficient, so I get it. I'm sure the general programming pattern is slow and speed matters here.

It's completely your call of course. You can fix it however you want, I'd just like to hit the `stop()` button =)

**Comment by rudiedirkx** [comment]
So how about that `stop()` button? Any way to let the listener stop the parser? Any way is fine with me.

**Comment by danyellnoe** [comment]
+1 I'd also like to see this implemented.  My workaround: once I've reached the limit of records (JSON objects) I'm consuming, I just stop processing the data coming from the parser.  That is, I just stop listening.  But, clearly this is inefficient.

**Comment by rudiedirkx** [comment]
@danyellnoe Throwing an excepting inside the listener is more efficient. It's a hack, but it stops.

**Comment by NigelRel3** [comment]
In the RegexListener example code example_regex.php, this is done in the order...
Create listener
Create Parser
Set listener closures for processing elements as something like 
```
$listener->setMatch(["/total_rows" => function ($data ) use ($parser) {
    echo "/total_rows=".$data.PHP_EOL;
    $parser->stop();
}]);
```
So this will fetch the /total_rows element (in the example) from ratherBig.json and then stop.

**Comment by MAXakaWIZARD** [comment]
Requested feature is implemented in v8.1.0: https://github.com/salsify/jsonstreamingparser/releases/tag/v8.1.0
